Hw03 l06 annotations

akuliniya opened the pull request:
Тестовый фреймворк

akuliniya committed:
ДЗ HW03-L06-annotations

akuliniya committed:
ДЗ HW03-L06-annotations

stvort commented on HW03-L06-annotations/src/main/java/homework/TestRunner.java:
```diff
@@ -0,0 +1,112 @@
+import java.lang.reflect.Method;
+
+public class TestRunner {
+    private final static Class<Before> BEFORE_ANNOTATION = Before.class;
+    private final static Class<After> AFTER_ANNOTATION = After.class;
+    private final static Class<Test> TEST_ANNOTATION = Test.class;
+
+    private static int tests = 0;
+    private static int passed = 0;
+    private static int failed = 0;
```
Лучше не использовать статические поля для передачи данных между методами. Они про состояние класса, общее для всех его пользователей. В данном случае это явно не так. Плюс это плохо для запуска из нескольких потоков. Эти поля могут быть локальными переменными, аргументами методов и возвращаемыми типами

stvort commented on HW03-L06-annotations/src/main/java/homework/TestRunner.java:
```diff
@@ -0,0 +1,112 @@
+                tests++;
+                try {
+                    Object test = constructor.newInstance();
+                    getBeforeMethod(methods).invoke(test);
```
Лучше заранее распределить методы по спискам и уже потом запускать тесты

stvort commented on HW03-L06-annotations/src/main/java/homework/TestRunner.java:
```diff
@@ -0,0 +1,112 @@
+                    getBeforeMethod(methods).invoke(test);
+                    method.invoke(test);
+                    passed++;
+                    getAfterMethod(methods).invoke(test);
```
`AfterMethod` лучше запускать даже в том случае, если тесты или `BeforeMethod` не прошел

akuliniya committed:
HW03-L06-annotations
1. Из TestRunner убраны статические поля
2. Методы заранее распределяются по спискам.
3. AfterMethod запускается даже в том случае, если тесты или BeforeMethod не прошел.

stvort commented on HW03-L06-annotations/src/main/java/homework/TestRunner.java:
```diff
@@ -4,109 +4,139 @@
-    private static void runTests(String testClassName) throws ClassNotFoundException, NoSuchMethodException, InstantiationException {
+    private static List<Integer> runTests(String testClassName) throws ClassNotFoundException, NoSuchMethodException, InstantiationException, InvocationTargetException, IllegalAccessException {
+        List<Integer> testRunResults = new ArrayList<>();
+        int testsCounter = 0;
+        int passedCounter = 0;
+        int failedCounter = 0;
```
Возможно нам нужен только один счетчик. Все остальное либо уже есть, либо вычисляется

stvort commented on HW03-L06-annotations/src/main/java/homework/TestRunner.java:
```diff
@@ -4,109 +4,139 @@
+            Object test = constructor.newInstance();
+            testsCounter++;
+
+            for (Method beforeMethod : beforeMethods) {
+                beforeMethod.invoke(test);
+            }
```
Если тут будет исключение, то after не выполнятся, а мы могли в before что-то важное успеть открыть, что нужно закрыть в after

akuliniya committed:
Merge branch 'main' into HW03-L06-annotations
